### ISSUE TITLE
Overload ambiguity is ignored in `self`-annotated methods

### ISSUE THREAD

**Issue by BvB93** [python/mypy#11347]
**Bug Report**

Per the title: mypy does not check for overload ambiguity in `self`-annotated methods.
Mypy will instead, effectively, ignore `self` and then return the first matching overload, 
thus circumventing the usual `Any` that is returned when overload ambiguity is involved.
This has lead to a number of false positives in https://github.com/numpy/numpy/issues/20099 and (in part) https://github.com/numpy/numpy/issues/20072. 

The conditions for triggering this bug are as following:

* The callable is a method (things work fine for functions).
* The method is overloaded.
* `self` must be annotated with a parametrized generic.

**To Reproduce**

``` python
from typing import Any, Generic, overload, TypeVar, TYPE_CHECKING
import numpy as np

_CharType = TypeVar("_CharType", np.str_, np.bytes_)


class CharArray(Generic[_CharType]):
    @overload
    def strip(self: CharArray[np.str_], chars: str = ...) -> CharArray[np.str_]: ...
    @overload
    def strip(self: CharArray[np.bytes_], chars: bytes = ...) -> CharArray[np.bytes_]: ...


ar_any: CharArray[Any]
ar_str: CharArray[np.str_]
ar_bytes: CharArray[np.bytes_]

if TYPE_CHECKING:
    # The good
    reveal_type(ar_str.strip())  # E: Revealed type is "__main__.CharArray[numpy.str_]"
    reveal_type(ar_bytes.strip())  # E: Revealed type is "__main__.CharArray[numpy.bytes_]"

    # The bad
    # "__main__.CharArray[Any]" would be expected here due to overload ambiguity
    reveal_type(ar_any.strip())  # E: Revealed type is "__main__.CharArray[numpy.str_]"

    # Calling the method directly from the class does do the trick
    reveal_type(CharArray.strip(ar_any))  # E: Revealed type is "test.CharArray[Any]"
```

**Your Environment**

- Mypy version used: 0.910
- Mypy command-line flags: n.a.
- Mypy configuration options from `mypy.ini` (and other config files): n.a.
- Python version used: 3.9.6
- Operating system and version: macOS 11.4

**Comment by sobolevn** [comment]
## Simplified reproduction

### Without `self` annotation

Let's see what will happen when we use regular `@overload` with `Any`

```python
from typing import Generic, TypeVar, overload, Any

T = TypeVar('T')

class Some:
    @overload
    def method(self, arg: int) -> bool: ...
    @overload
    def method(self, arg: str) -> float: ...
    def method(self): ...

# ok
s1: Some
s2: Some
reveal_type(s1.method(1))  # N: Revealed type is "builtins.bool"
reveal_type(s2.method('a')) # N: Revealed type is "builtins.float"

# also ok
s3: Some
a: Any
reveal_type(s3.method(a))  # N: Revealed type is "Any"
```

### With `self` annotation

```python
from typing import Generic, TypeVar, overload, Any

T = TypeVar('T')

class Some(Generic[T]):
    @overload
    def method(self: Some[int]) -> bool: ...
    @overload
    def method(self: Some[str]) -> float: ...
    def method(self): ...

# ok
s1: Some[int]
s2: Some[str]
reveal_type(s1.method())  # N: Revealed type is "builtins.bool"
reveal_type(s2.method())  # N: Revealed type is "builtins.float"

# error
s3: Some[Any]
reveal_type(s3.method()) # N: Revealed type is "builtins.bool"
# Should be: Revealed type is "Any"
```

## Proposed solution

Wrong `@overload` cases should also be `Any` even with `self` annotations.

I will send a PR with the fix.
